File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotHardcodeCredentials_DefaultValues.cs
using System;
using System.Net;
using System.Security;
using System.Security.Cryptography;

namespace Tests.Diagnostics
{
    class Program
    {
        public const string DBConnectionString = "Server=localhost; Database=Test; User=SA; Password=Secret123";    // Noncompliant

        private const string secret = "constantValue";

        public void Test(string user)
        {
            string password = @"foo"; // Noncompliant {{"password" detected here, make sure this is not a hard-coded credential.}}
//                 ^^^^^^^^^^^^^^^^^

            string foo, passwd = "a"; // Noncompliant {{"passwd" detected here, make sure this is not a hard-coded credential.}}
//                      ^^^^^^^^^^^^

            string pwdPassword = "a"; // Noncompliant {{"pwd, password" detected here, make sure this is not a hard-coded credential.}}

            string foo2 = @"Password=123"; // Noncompliant

            string bar;
            bar = "Password=p"; // Noncompliant
//          ^^^^^^^^^^^^^^^^^^

            foo = "password=";
            foo = "passwordpassword";
            foo = "foo=1;password=1"; // Noncompliant
            foo = "foo=1password=1";
            foo = ""; // Compliant

            string myPassword1 = null;
            string myPassword2 = "";
            string myPassword3 = "        ";
            string myPassword4 = @"foo"; // Noncompliant
            string query2 = "password=hardcoded;user='" + user + "'"; // Noncompliant
        }

        public void DefaultKeywords()
        {
            string password = "a";       // Noncompliant
            string x1 = "password=a";    // Noncompliant

            string passwd = "a";         // Noncompliant
            string x2 = "passwd=a";      // Noncompliant

            string pwd = "a";            // Noncompliant
            string x3 = "pwd=a";         // Noncompliant

            string passphrase = "a";     // Noncompliant
            string x4 = "passphrase=a";  // Noncompliant
        }

        public void Constants()
        {
            const string ConnectionString = "Server=localhost; Database=Test; User=SA; Password=Secret123";    // Noncompliant
            const string ConnectionStringWithSpaces = "Server=localhost; Database=Test; User=SA; Password   =   Secret123";    // Noncompliant
            const string Password = "Secret123";  // Noncompliant

            const string LoginName = "Admin";
            const string Localhost = "localhost";
        }

        public void StandardAPI(SecureString secureString, string nonHardcodedPassword, byte[] byteArray, CspParameters cspParams)
        {
            var networkCredential = new NetworkCredential();
            networkCredential.Password = nonHardcodedPassword;
            networkCredential.Domain = "hardcodedDomain";
            new NetworkCredential("username", secureString);
            new NetworkCredential("username", nonHardcodedPassword);
            new NetworkCredential("username", secureString, "domain");
            new NetworkCredential("username", nonHardcodedPassword, "domain");

            new PasswordDeriveBytes(nonHardcodedPassword, byteArray);
            new PasswordDeriveBytes(new byte[] { 1 }, byteArray);
            new PasswordDeriveBytes(nonHardcodedPassword, byteArray, cspParams);
            new PasswordDeriveBytes(new byte[] { 1 }, byteArray, cspParams);
            new PasswordDeriveBytes(nonHardcodedPassword, byteArray, "strHashName", 1);
            new PasswordDeriveBytes(new byte[] { 1 }, byteArray, "strHashName", 1);
            new PasswordDeriveBytes(nonHardcodedPassword, byteArray, "strHashName", 1, cspParams);
            new PasswordDeriveBytes(new byte[] { 1 }, byteArray, "strHashName", 1, cspParams);

            new NetworkCredential("username", secret); // Noncompliant
            new NetworkCredential("username", "hardcoded"); // Noncompliant
            new NetworkCredential("username", "hardcoded", "domain"); // Noncompliant
            networkCredential.Password = "hardcoded"; // Noncompliant
            new PasswordDeriveBytes("hardcoded", byteArray); // Noncompliant
            new PasswordDeriveBytes("hardcoded", byteArray, cspParams); // Noncompliant
            new PasswordDeriveBytes("hardcoded", byteArray, "strHashName", 1); // Noncompliant
            new PasswordDeriveBytes("hardcoded", byteArray, "strHashName", 1, cspParams); // Noncompliant
        }

        public void CompliantParameterUse(string pwd)
        {
            string query1 = "password=?";
            string query2 = "password=:password";
            string query3 = "password=:param";
            string query4 = "password='" + pwd + "'";
            string query5 = "password={0}";
            string query6 = "password=;user=;";
            string query7 = "password=:password;user=:user;";
            string query8 = "password=?;user=?;";
            string query9 = @"Server=myServerName\myInstanceName;Database=myDataBase;Password=:myPassword;User Id=:username;";
        }

        public void WordInVariableNameAndValue()
        {
            // It's compliant when the word is used in name AND the value.
            const string PASSWORD = "Password";
            const string Password_Input = "[id='password']";
            const string PASSWORD_PROPERTY = "custom.password";
            const string TRUSTSTORE_PASSWORD = "trustStorePassword";
            const string CONNECTION_PASSWORD = "connection.password";
            const string RESET_PASSWORD = "/users/resetUserPassword";
            const string RESET_PASSWORD_CS = "/uzivatel/resetovat-heslo"; // Noncompliant, "heslo" means "password", but we don't translate SEO friendly URL for all languages

            string passwordKey = "Password";
            string passwordProperty = "config.password.value";
            string passwordName = "UserPasswordValue";
            string password = "Password";
            string pwd = "pwd";

            string myPassword = "pwd"; // Noncompliant, different value from word list is used
        }

        public void UriWithUserInfo(string pwd, string domain)
        {
            string n1 = "scheme://user:azerty123@domain.com"; // Noncompliant {{Review this hard-coded URI, which may contain a credential.}}
            string n2 = "scheme://user:With%20%3F%20Encoded@domain.com";              // Noncompliant
            string n3 = "scheme://user:With!$&'()*+,;=OtherCharacters@domain.com";    // Noncompliant
            string n4 = "scheme://user:azerty123@" + domain;  // Noncompliant

            string c1 = "scheme://user:" + pwd + "@domain.com";
            string c2 = "scheme://user:@domain.com";
            string c3 = "scheme://user@domain.com:80";
            string c4 = "scheme://user@domain.com";
            string c5 = "scheme://domain.com/user:azerty123";
            string c6 = String.Format("scheme://user:{0}@domain.com", pwd);
            string c7 = $"scheme://user:{pwd}@domain.com";

            string e1 = "scheme://admin:admin@domain.com";    // Compliant exception, user and password are the same
            string e2 = "scheme://abc:abc@domain.com";        // Compliant exception, user and password are the same
            string e3 = "scheme://a%20;c:a%20;c@domain.com";  // Compliant exception, user and password are the same
        }

        public void LiteralAsArgument(string pwd, string server)
        {
            using (var conn = new SqlConnection("Server = localhost; Database = Test; User = SA; Password = Secret123")) { } // Noncompliant
            using (var conn = OpenConn("Server = localhost; Database = Test; User = SA; Password = Secret123")) { } // Noncompliant
            using (var conn = OpenConn("Server = " + server + "; Database = Test; User = SA; Password = Secret123")) { } // Noncompliant

            using (var conn = OpenConn("Server = localhost; Database = Test; User = SA; Password = " + pwd)) { }
        }

        private SqlConnection OpenConn(string connectionString)
        {
            var ret = new SqlConnection(connectionString);
            ret.Open();
            return ret;
        }

        public string SecretConnectionStringProperty
        {
            get
            {
                return "Server = localhost; Database = Test; User = SA; Password = Secret123"; // Noncompliant
            }
        }

        public string SecretConnectionStringProperty_OK
        {
            get
            {
                return "Nothing to see here";
            }
        }

        public string SecretConnectionStringProperty2 => "Server = localhost; Database = Test; User = SA; Password = Secret123"; // Noncompliant
        public string SecretConnectionStringProperty2_OK => "Nothing to see here";

        public string SecretConnectionStringProperty3 { get; } = "Server = localhost; Database = Test; User = SA; Password = Secret123"; // Noncompliant
        public string SecretConnectionStringProperty3_OK { get; } = "Nothing to see here";

        public string SecretConnectionStringFunction()
        {
            return "Server = localhost; Database = Test; User = SA; Password = Secret123"; // Noncompliant
        }

        public string SecretConnectionStringFunction_OK()
        {
            return "Nothing to see here";
        }

        public string SecretConnectionStringFunction2() => "Server = localhost; Database = Test; User = SA; Password = Secret123"; // Noncompliant
        public string SecretConnectionStringFunction2_OK() => "Nothing to see here";
    }

    class SqlConnection : IDisposable
    {
        public SqlConnection(string connectionString) { }
        public void Open() { }
        public void Dispose() { }
    }

    class FalseNegatives
    {
        private string password;

        public void Foo(string user)
        {
            this.password = "foo"; // False Negative
            Configuration.Password = "foo"; // False Negative
            this.password = Configuration.Password = "foo"; // False Negative
            string query1 = "password=':crazy;secret';user=xxx"; // False Negative - passwords enclosed in '' are not covered
        }

        class Configuration
        {
            public static string Password { get; set; }
        }
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotHardcodeCredentials_DefaultValues.vb
Imports System
Imports System.Net
Imports System.Security
Imports System.Security.Cryptography

Namespace Tests.Diagnostics

    Class Program

        Public Const DBConnectionString As String = "Server=localhost; Database=Test; User=SA; Password=Secret123"    ' Noncompliant
        Public Const EditPasswordPageUrlToken As String = "{Account.EditPassword.PageUrl}" ' Compliant

        Private Const Secret As String = "constantValue"

        Public Sub Test(User As String)
            Dim password As String = "foo" 'Noncompliant {{"password" detected here, make sure this is not a hard-coded credential.}}
            '   ^^^^^^^^^^^^^^^^^^^^^^^^^^
            Dim foo As String, passwd As String = "a" 'Noncompliant {{"passwd" detected here, make sure this is not a hard-coded credential.}}
            '                  ^^^^^^^^^^^^^^^^^^^^^^
            Dim pwdPassword As String = "a"     'Noncompliant {{"pwd, password" detected here, make sure this is not a hard-coded credential.}}
            Dim foo2 As String = "Password=123" 'Noncompliant
            Dim bar As String
            bar = "Password=p" 'Noncompliant ^13#18

            foo = "password"
            foo = "password="
            foo = "passwordpassword"
            foo = "foo=1;password=1" 'Noncompliant
            foo = "foo=1password=1"

            Dim myPassword1 As String = Nothing
            Dim myPassword2 As String = ""
            Dim myPassword3 As String = "   "
            Dim myPassword4 As String = "foo" 'Noncompliant
            Dim query2 As String = "password=hardcoded;user='" + User + "'" ' Noncompliant
        End Sub

        Public Sub DefaultKeywords()
            Dim password As String = "a"       ' Noncompliant
            Dim x1 As String = "password=a"    ' Noncompliant

            Dim passwd As String = "a"         ' Noncompliant
            Dim x2 As String = "passwd=a"      ' Noncompliant

            Dim pwd As String = "a"            ' Noncompliant
            Dim x3 As String = "pwd=a"         ' Noncompliant

            Dim passphrase As String = "a"     ' Noncompliant
            Dim x4 As String = "passphrase=a"  ' Noncompliant
        End Sub

        Public Sub Constants()
            Const ConnectionString As String = "Server=localhost; Database=Test; User=SA; Password=Secret123"    ' Noncompliant
            Const ConnectionStringWithSpaces As String = "Server=localhost; Database=Test; User=SA; Password   =   Secret123"    ' Noncompliant
            Const Password As String = "Secret123"  ' Noncompliant

            Const LoginName As String = "Admin"
            Const Localhost As String = "localhost"
        End Sub

        Public Sub StandardAPI(secureString As SecureString, nonHardcodedPassword As String, byteArray As Byte(), cspParams As CspParameters)
            Dim networkCredential = New NetworkCredential()
            networkCredential.Password = nonHardcodedPassword
            networkCredential.Domain = "hardcodedDomain"
            networkCredential = New NetworkCredential("username", secureString)
            networkCredential = New NetworkCredential("username", nonHardcodedPassword)
            networkCredential = New NetworkCredential("username", secureString, "domain")
            networkCredential = New NetworkCredential("username", nonHardcodedPassword, "domain")

            Dim passwordDeriveBytes = New PasswordDeriveBytes(nonHardcodedPassword, byteArray)
            passwordDeriveBytes = New PasswordDeriveBytes(New Byte() {1}, byteArray)
            passwordDeriveBytes = New PasswordDeriveBytes(nonHardcodedPassword, byteArray, cspParams)
            passwordDeriveBytes = New PasswordDeriveBytes(New Byte() {1}, byteArray, cspParams)
            passwordDeriveBytes = New PasswordDeriveBytes(nonHardcodedPassword, byteArray, "strHashName", 1)
            passwordDeriveBytes = New PasswordDeriveBytes(New Byte() {1}, byteArray, "strHashName", 1)
            passwordDeriveBytes = New PasswordDeriveBytes(nonHardcodedPassword, byteArray, "strHashName", 1, cspParams)
            passwordDeriveBytes = New PasswordDeriveBytes(New Byte() {1}, byteArray, "strHashName", 1, cspParams)

            networkCredential = New NetworkCredential("username", Secret) 'Noncompliant {{Please review this hard-coded password.}}
            networkCredential = New NetworkCredential("username", "hardcoded") 'Noncompliant {{Please review this hard-coded password.}}
            networkCredential = New NetworkCredential("username", "hardcoded", "domain") 'Noncompliant {{Please review this hard-coded password.}}
            networkCredential.Password = "hardcoded" 'Noncompliant {{Please review this hard-coded password.}}
            passwordDeriveBytes = New PasswordDeriveBytes("hardcoded", byteArray) 'Noncompliant {{Please review this hard-coded password.}}
            passwordDeriveBytes = New PasswordDeriveBytes("hardcoded", byteArray, cspParams) 'Noncompliant {{Please review this hard-coded password.}}
            passwordDeriveBytes = New PasswordDeriveBytes("hardcoded", byteArray, "strHashName", 1) 'Noncompliant {{Please review this hard-coded password.}}
            passwordDeriveBytes = New PasswordDeriveBytes("hardcoded", byteArray, "strHashName", 1, cspParams) 'Noncompliant {{Please review this hard-coded password.}}
        End Sub

        Public Sub CompliantParameterUse(pwd As String)
            Dim query1 As String = "password=?"
            Dim query2 As String = "password=:password"
            Dim query3 As String = "password=:param"
            Dim query4 As String = "password='" + pwd + "'"
            Dim query5 As String = "password='" & pwd & "'"
            Dim query6 As String = "password={0}"
            Dim query7 As String = "password=;user=;"
            Dim query8 As String = "password=:password;user=:user;"
            Dim query9 As String = "password=?;user=?;"
            Dim query10 As String = "Server=myServerName\myInstanceName;Database=myDataBase;Password=:myPassword;User Id=:username;"
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = ?")
            End Using
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = :password")
            End Using
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = {0}")
            End Using
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = ")
            End Using
        End Sub

        Public Sub WordInConstantNameAndValue()
            ' It's compliant when the word is used in name AND the value.
            Const PASSWORD As String = "Password"
            Const Password_Input As String = "[id='password']"
            Const PASSWORD_PROPERTY As String = "custom.password"
            Const TRUSTSTORE_PASSWORD As String = "trustStorePassword"
            Const CONNECTION_PASSWORD As String = "connection.password"
            Const RESET_PASSWORD As String = "/users/resetUserPassword"
            Const RESET_PASSWORD_CS As String = "/uzivatel/resetovat-heslo" ' Noncompliant, "heslo" means "password", but we don't translate SEO friendly URL for all languages
        End Sub

        Public Sub WordInVariableNameAndValue()
            ' It's compliant when the word is used in name AND the value.
            Dim PasswordKey As String = "Password"
            Dim PasswordProperty As String = "config.password.value"
            Dim PasswordName As String = "UserPasswordValue"
            Dim Password As String = "Password"
            Dim pwd As String = "pwd"

            Dim myPassword As String = "pwd"    ' Noncompliant, different value from word list is used
        End Sub

        Public Sub UriWithUserInfo(Pwd As String, Domain As String)
            Dim n1 As String = "scheme://user:azerty123@domain.com" ' Noncompliant {{Review this hard-coded URI, which may contain a credential.}}
            Dim n2 As String = "scheme://user:With%20%3F%20Encoded@domain.com"              ' Noncompliant
            Dim n3 As String = "scheme://user:With!$&'()*+,;=OtherCharacters@domain.com"    ' Noncompliant
            Dim n4 As String = "scheme://user:azerty123@" + Domain  ' Noncompliant
            Dim n5 As String = "scheme://user:azerty123@" & Domain  ' Noncompliant

            Dim c1 As String = "scheme://user:" + Pwd + "@domain.com"
            Dim c2 As String = "scheme://user:" & Pwd & "@domain.com"
            Dim c3 As String = "scheme://user:@domain.com"
            Dim c4 As String = "scheme://user@domain.com:80"
            Dim c5 As String = "scheme://user@domain.com"
            Dim c6 As String = "scheme://domain.com/user:azerty123"
            Dim c7 As String = String.Format("scheme://user:{0}@domain.com", Pwd)
            Dim c8 As String = $"scheme://user:{Pwd}@domain.com"

            Dim e1 As String = "scheme://admin:admin@domain.com"    ' Compliant exception, user and password are the same
            Dim e2 As String = "scheme://abc:abc@domain.com"        ' Compliant exception, user and password are the same
            Dim e3 As String = "scheme://a%20;c:a%20;c@domain.com"  ' Compliant exception, user and password are the same

            Dim html1 As String = ' Noncompliant
"This is article http://login:secret@www.example.com
Email: info@example.com
Phone: +0000000"

            Dim html2 As String =
"This is article http://www.example.com
Email: info@example.com
Phone: +0000000"

            Dim html3 As String = "This is article http://www.example.com Email: info@example.com Phone: +0000000"
            Dim html4 As String = "This is article http://www.example.com<br>Email:info@example.com<br>Phone:+0000000"
            Dim html5 As String = "This is article http://user:secret@www.example.com<br>Email:info@example.com<br>Phone:+0000000" ' Noncompliant
        End Sub

        Public Sub LiteralAsArgument(pwd As String, server As String)
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = Secret123")  ' Noncompliant
            End Using
            Using Conn As SqlConnection = OpenConn("Server = localhost; Database = Test; User = SA; Password = Secret123") ' Noncompliant
            End Using
            Using Conn As New SqlConnection("Server = " + server + "; Database = Test; User = SA; Password = Secret123") ' Noncompliant
            End Using
            Using Conn As New SqlConnection("Server = " & server & "; Database = Test; User = SA; Password = Secret123") ' Noncompliant
            End Using

            Using OpenConn("password")
            End Using
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = " + pwd)
            End Using
            Using Conn As New SqlConnection("Server = localhost; Database = Test; User = SA; Password = " & pwd)
            End Using
        End Sub

        Private Function OpenConn(connectionString As String) As SqlConnection
            Dim Ret As New SqlConnection(connectionString)
            Ret.Open()
            Return Ret
        End Function

        Public ReadOnly Property ConnectionStringProperty As String
            Get
                Return "Server = localhost; Database = Test; User = SA; Password = Secret123" ' Noncompliant
            End Get
        End Property

        Public ReadOnly Property ConnectionStringProperty_OK As String
            Get
                Return "Nothing to see here"
            End Get
        End Property

        Public ReadOnly Property ConnectionStringProperty2 As String = "Server = localhost; Database = Test; User = SA; Password = Secret123" ' Noncompliant
        Public ReadOnly Property ConnectionStringProperty2_OK As String = "Nothing to see here"

        Public Function ConnectionStringFunction() As String
            Return "Server = localhost; Database = Test; User = SA; Password = Secret123" ' Noncompliant
        End Function

        Public Function ConnectionStringFunction_OK() As String
            Return "Nothing to see here"
        End Function

    End Class

    Public Class SqlConnection
        Implements IDisposable

        Public Sub New(ConnStr As String)
        End Sub

        Public Sub Open()
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
        End Sub

    End Class

    Class FalseNegatives
        Private password As String

        Public Sub Foo(user As String)
            Me.password = "foo" ' False Negative
            Configuration.Password = "foo" ' False Negative
            Me.password = Configuration.Password = "foo" ' False Negative
            Dim query1 As String = "password=':crazy;secret';user=xxx" ' False Negative - passwords enclosed in '' are not covered
        End Sub

        Class Configuration
            Public Shared Property Password As String
        End Class
    End Class

End Namespace
